src/cargo/core/compiler/compilation.rs

```diff
@@ -178,16 +178,17 @@ impl<'cfg> Compilation<'cfg> {
             search_path
         };
 
-        search_path.extend(util::dylib_path().into_iter());
-        if cfg!(target_os = "macos") {
+        let dylib_path = util::dylib_path();
+        let dylib_path_is_empty = dylib_path.is_empty();
+        search_path.extend(dylib_path.into_iter());
+        if cfg!(target_os = "macos") && dylib_path_is_empty {
             // These are the defaults when DYLD_FALLBACK_LIBRARY_PATH isn't
-            // set. Since Cargo is explicitly setting the value, make sure the
-            // defaults still work.
-            if let Ok(home) = env::var("HOME") {
+            // set or set to an empty string. Since Cargo is explicitly setting
+            // the value, make sure the defaults still work.
+            if let Some(home) = env::var_os("HOME") {
                 search_path.push(PathBuf::from(home).join("lib"));
             }
             search_path.push(PathBuf::from("/usr/local/lib"));
-            search_path.push(PathBuf::from("/lib"));
```

Comment:
Why remove `/lib`?

Reply:
It's explained in the corresponding commit message. In short, contrary to what the dyld documentation says, it's not in the default, and never has been (and the doc has always been wrong).